### DISCUSSION TITLE
Question about extension concurrency

### DISCUSSION THREAD

**Discussion by vassilvk** [santhosh-tekuri/jsonschema#152] [General]
Given an extension for a keyword, if the keyword is found multiple times in a document, the extension schema's `Validate` method will be invoked once per keyword occurrence.

The question is, will those `Validate` calls be made concurrently? If the extension is used to accumulate state from within `Validate`, should one protect against race conditions?

**Comment by santhosh-tekuri** [comment]
there are no go routines created during `Compile\Validate`. so no need to protect against race conditions

but note that user may call validate from multiple go-routines.

I am not sure where do you want to maintain state. if you are planing to store in `ExtSchema` implementation. please do not do that. since users may be calling `schema.validate` from multiple routines.

currently I don't think the extension api has any provision to store state within single schema validate call

**Reply by vassilvk** [reply]
Thank you @santhosh-tekuri,

> but note that user may call validate from multiple go-routine

Understood. In my use case I am in control of calls to the top `(*Schema).Validate` for a given compiled schema, so there's no scenario in which `(*Schema).Validate` itself will be called concurrently. The question was about concurrency on calls to `(*ExtSchema).Validate` spawned from a single call to `(*Schema).Validate`.

Thanks for your prompt answer :)

**Reply by santhosh-tekuri** [reply]
There is no concurrency used in the library
